url47e6aca666_pocorall_automaton_tgz-pJ8-net_pocorall_automaton_DatatypesJ8/src/net/pocorall/automaton/BasicOperations.java

```diff
@@ -0,0 +1,622 @@
+		worklist.add(p);
+		visited.add(p);
+		while (worklist.size() > 0) {
+			p = worklist.removeFirst();
+			if (p.s1.accept != null && p.s2.accept == null)
+				return false;
+			Transition[] t1 = transitions1[p.s1.number];
+			Transition[] t2 = transitions2[p.s2.number];
+			for (int n1 = 0, b2 = 0; n1 < t1.length; n1++) {
+				while (b2 < t2.length && t2[b2].max < t1[n1].min)
+					b2++;
+				int min1 = t1[n1].min, max1 = t1[n1].max;
+				for (int n2 = b2; n2 < t2.length && t1[n1].max >= t2[n2].min; n2++) {
+					if (t2[n2].min > min1)
+						return false;
+					if (t2[n2].max < Character.MAX_VALUE)
+						min1 = t2[n2].max + 1;
+					else {
+						min1 = Character.MAX_VALUE;
+						max1 = Character.MIN_VALUE;
+					}
+					StatePair q = new StatePair(t1[n1].to, t2[n2].to);
+					if (!visited.contains(q)) {
+						worklist.add(q);
+						visited.add(q);
+					}
+				}
+				if (min1 <= max1)
+					return false;
+			}
+		}
+		return true;
+	}
+
+	/**
+	 * Returns an automaton that accepts the union of the languages of the given automata.
+	 * <p/>
+	 * Complexity: linear in number of states.
+	 */
+	public static DefaultAutomaton union(DefaultAutomaton a1, DefaultAutomaton a2) {
+		if ((a1.isSingleton() && a2.isSingleton() && a1.singleton.equals(a2.singleton)) || a1 == a2)
+			return a1.cloneIfRequired();
+		if (a1 == a2) {
+			a1 = a1.cloneExpanded();
+			a2 = a2.cloneExpanded();
+		} else {
+			a1 = a1.cloneExpandedIfRequired();
+			a2 = a2.cloneExpandedIfRequired();
+		}
+		State s = new State();
+		s.addEpsilon(a1.initial);
+		s.addEpsilon(a2.initial);
+		a1.initial = s;
+		a1.deterministic = false;
+		a1.clearHashCode();
+		a1.checkMinimizeAlways();
+		return a1;
+	}
+
+	/**
+	 * Returns an automaton that accepts the union of the languages of the given automata.
+	 * <p/>
+	 * Complexity: linear in number of states.
+	 */
+	public static DefaultAutomaton union(Collection<DefaultAutomaton> l) {
+		Set<Integer> ids = new HashSet<Integer>();
+		for (DefaultAutomaton a : l)
+			ids.add(System.identityHashCode(a));
+		boolean has_aliases = ids.size() != l.size();
+		State s = new State();
+		for (DefaultAutomaton b : l) {
+			if (b.isEmpty())
+				continue;
+			DefaultAutomaton bb = b;
+			if (has_aliases)
+				bb = bb.cloneExpanded();
+			else
+				bb = bb.cloneExpandedIfRequired();
+			s.addEpsilon(bb.initial);
+		}
+		DefaultAutomaton a = new DefaultAutomaton();
+		a.initial = s;
+		a.deterministic = false;
+		a.clearHashCode();
+		a.checkMinimizeAlways();
+		return a;
+	}
+
+	/**
+	 * Determinizes the given automaton.
+	 * <p/>
+	 * Complexity: exponential in number of states.
+	 */
+	public static void determinize(LinkedAutomaton a) {
+		if (a.deterministic)
+			return;
+		Set<State> initialset = new HashSet<State>();
+		initialset.add(a.initial);
+		determinize(a, initialset);
+	}
+
+	/**
+	 * Determinizes the given automaton using the given set of initial states.
+	 */
+	static void determinize(LinkedAutomaton a, Set<State> initialset) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.